File: src/main/java/rars/riscv/hardware/MemoryConfiguration.java
package rars.riscv.hardware;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public enum MemoryConfiguration {
    DEFAULT(
            "Default",
            "Default",
            0x00400000, // .text Base Address
            0x10000000, // Data Segment base address
            0x10000000, // .extern Base Address
            0x10008000, // Global Pointer $gp)
            0x10010000, // .data base Address
            0x10040000, // heap base address
            0x7fffeffc, // stack pointer $sp (from SPIM not MIPS)
            0x7ffffffc, // stack base address
            0x7fffffff, // highest address in user space
            0x80000000, // lowest address in kernel space
            0xffff0000, // MMIO base address
            0xffffffff, // highest address in kernel (and memory)
            0x7fffffff, // data segment limit address
            0x0ffffffc, // text limit address
            0x10040000, // stack limit address
            0xffffffff // memory map limit address
    ),
    DATA_BASED_COMPACT(
            "CompactDataAtZero",
            "Compact, data at address 0",
            0x00003000, // .text Base Address
            0x00000000, // Data Segment base address
            0x00001000, // .extern Base Address
            0x00001800, // Global Pointer $gp)
            0x00000000, // .data base Address
            0x00002000, // heap base address
            0x00002ffc, // stack pointer $sp
            0x00002ffc, // stack base address
            0x00003fff, // highest address in user space
            0x00004000, // lowest address in kernel space
            0x00007f00, // MMIO base address
            0x00007fff, // highest address in kernel (and memory)
            0x00002fff, // data segment limit address
            0x00003ffc, // text limit address
            0x00002000, // stack limit address
            0x00007fff // memory map limit address
    ),
    TEXT_BASED_COMPACT(
            "CompactTextAtZero",
            "Compact, text at address 0",
            0x00000000, // .text Base Address
            0x00001000, // Data Segment base address
            0x00001000, // .extern Base Address
            0x00001800, // Global Pointer $gp)
            0x00002000, // .data base Address
            0x00003000, // heap base address
            0x00003ffc, // stack pointer $sp
            0x00003ffc, // stack base address
            0x00003fff, // highest address in user space
            0x00004000, // lowest address in kernel space
            0x00007f00, // MMIO base address
            0x00007fff, // highest address in kernel (and memory)
            0x00003fff, // data segment limit address
            0x00000ffc, // text limit address
            0x00003000, // stack limit address
            0x00007fff // memory map limit address
    );

    public final @NotNull String identifier, description;
    /** base address for (user) text segment */
    public final int textBaseAddress;
    /** base address for (user) data segment */
    public final int dataSegmentBaseAddress;
    /** base address for .extern directive */
    public final int externBaseAddress;
    /** base address for storing globals */
    public final int globalPointerAddress;
    /** base address for storage of non-global static data in data segment */
    public final int dataBaseAddress;
    /** base address for heap */
    public final int heapBaseAddress;
    /** starting address for stack */
    public final int stackPointerAddress;
    /** base address for stack */
    public final int stackBaseAddress;
    /** highest address accessible in user (not kernel) mode. */
    public final int userHighAddress;
    /** kernel boundary. Only OS can access this or higher address */
    public final int kernelBaseAddress;
    /**  starting address for memory mapped I/O */
    public final int memoryMapBaseAddress;
    /** highest address acessible in kernel mode. */
    public final int kernelHighAddress;
    public final int dataSegmentLimitAddress;
    public final int textLimitAddress;
    public final int stackLimitAddress;
    public final int memoryMapLimitAddress;
    MemoryConfiguration(
            @NotNull final String identifier,
            @NotNull final String description,
            final int textBaseAddress,
            final int dataSegmentBaseAddress,
            final int externBaseAddress,
            final int globalPointerAddress,
            final int dataBaseAddress,
            final int heapBaseAddress,
            final int stackPointerAddress,
            final int stackBaseAddress,
            final int userHighAddress,
            final int kernelBaseAddress,
            final int memoryMapBaseAddress,
            final int kernelHighAddress,
            final int dataSegmentLimitAddress,
            final int textLimitAddress,
            final int stackLimitAddress,
            final int memoryMapLimitAddress) {
        this.identifier = identifier;
        this.description = description;
        this.textBaseAddress = textBaseAddress;
        this.dataSegmentBaseAddress = dataSegmentBaseAddress;
        this.externBaseAddress = externBaseAddress;
        this.globalPointerAddress = globalPointerAddress;
        this.dataBaseAddress = dataBaseAddress;
        this.heapBaseAddress = heapBaseAddress;
        this.stackPointerAddress = stackPointerAddress;
        this.stackBaseAddress = stackBaseAddress;
        this.userHighAddress = userHighAddress;
        this.kernelBaseAddress = kernelBaseAddress;
        this.memoryMapBaseAddress = memoryMapBaseAddress;
        this.kernelHighAddress = kernelHighAddress;
        this.dataSegmentLimitAddress = dataSegmentLimitAddress;
        this.textLimitAddress = textLimitAddress;
        this.stackLimitAddress = stackLimitAddress;
        this.memoryMapLimitAddress = memoryMapLimitAddress;
    }

    public static @Nullable MemoryConfiguration fromIdString(final @NotNull String id) {
        for (final MemoryConfiguration configuration : MemoryConfiguration.values()) {
            if (configuration.identifier.equals(id)) {
                return configuration;
            }
        }
        return null;
    }
}

File: src/main/java/rars/riscv/hardware/MemoryConfiguration.kt
package rars.riscv.hardware

enum class MemoryConfiguration(
    val identifier: String,
    @JvmField val description: String,
    /** base address for (user) text segment  */
    @JvmField val textBaseAddress: Int,
    /** base address for (user) data segment  */
    @JvmField val dataSegmentBaseAddress: Int,
    /** base address for .extern directive  */
    @JvmField val externBaseAddress: Int,
    /** base address for storing globals  */
    @JvmField val globalPointerAddress: Int,
    /** base address for storage of non-global static data in data segment  */
    @JvmField val dataBaseAddress: Int,
    /** base address for heap  */
    @JvmField val heapBaseAddress: Int,
    /** base address for stack  */
    @JvmField val stackBaseAddress: Int,
    /** highest address accessible in user (not kernel) mode.  */
    @JvmField val userHighAddress: Int,
    /** kernel boundary. Only OS can access this or higher address  */
    @JvmField val kernelBaseAddress: Int,
    /** starting address for memory mapped I/O  */
    @JvmField val memoryMapBaseAddress: Int,
    /** highest address acessible in kernel mode.  */
    @JvmField val kernelHighAddress: Int,
    @JvmField val dataSegmentLimitAddress: Int,
    @JvmField val textLimitAddress: Int,
    @JvmField val stackLimitAddress: Int,
    @JvmField val memoryMapLimitAddress: Int
) {
    DEFAULT(
        "Default",
        "Default",
        textBaseAddress = 0x00400000,
        textLimitAddress = 0x0ffffffc,

        dataSegmentBaseAddress = 0x10000000,
        externBaseAddress = 0x10000000,

        globalPointerAddress = 0x10008000,
        dataBaseAddress = 0x10010000,

        heapBaseAddress = 0x10040000,
        stackLimitAddress = 0x10040000,
        stackBaseAddress = 0x7ffffffc,

        userHighAddress = 0x7fffffff,
        dataSegmentLimitAddress = 0x7fffffff,

        kernelBaseAddress = -0x80000000,
        memoryMapBaseAddress = -0x00010000,
        memoryMapLimitAddress = -0x00000001,
        kernelHighAddress = -0x00000001,
    ),
    DATA_BASED_COMPACT(
        "CompactDataAtZero",
        "Compact, data at address 0",

        dataSegmentBaseAddress = 0x00000000,
        externBaseAddress = 0x00001000,

        globalPointerAddress = 0x00001800,
        dataBaseAddress = 0x00000000,
        heapBaseAddress = 0x00002000,
        stackBaseAddress = 0x00002ffc,
        dataSegmentLimitAddress = 0x00002fff,
        stackLimitAddress = 0x00002000,

        textBaseAddress = 0x00003000,
        textLimitAddress = 0x00003ffc,
        userHighAddress = 0x00003fff,

        kernelBaseAddress = 0x00004000,
        memoryMapBaseAddress = 0x00007f00,
        memoryMapLimitAddress = 0x00007fff,
        kernelHighAddress = 0x00007fff,
    ),
    TEXT_BASED_COMPACT(
        "CompactTextAtZero",
        "Compact, text at address 0",

        textBaseAddress = 0x00000000,
        textLimitAddress = 0x00000ffc,

        dataSegmentBaseAddress = 0x00001000,
        externBaseAddress = 0x00001000,

        globalPointerAddress = 0x00001800,
        dataBaseAddress = 0x00002000,

        heapBaseAddress = 0x00003000,
        stackLimitAddress = 0x00003000,
        stackBaseAddress = 0x00003ffc,

        userHighAddress = 0x00003fff,
        dataSegmentLimitAddress = 0x00003fff,

        kernelBaseAddress = 0x00004000,
        memoryMapBaseAddress = 0x00007f00,
        memoryMapLimitAddress = 0x00007fff,
        kernelHighAddress = 0x00007fff,
    );

    @JvmField
    val stackPointerAddress: Int = stackBaseAddress.alignTo(4)

    override fun toString() = identifier

    companion object {
        @JvmStatic
        fun fromIdString(id: String) = entries.find { it.identifier == id }
    }
}

/**
 * Represents a memory configuration of the given memory object.
 */

infix fun Int.alignTo(alignment: Int) = this / alignment * alignment
infix fun Long.alignTo(alignment: Long) = this / alignment * alignment
infix fun Short.alignTo(alignment: Short) = this / alignment * alignment
infix fun Byte.alignTo(alignment: Byte) = this / alignment * alignment

//infix fun 